### PR TITLE
Add a new "ScanForModuleDependencies" flag.

### PR DESCRIPTION
**Pull request by LoneBoco**
**What does this PR do?**

This controls the Visual Studio "Scan Sources for Module Dependencies" option, which controls how it scans for C++20 modules.  This new flag sets the option in the generated project files.

**How does this PR change Premake's behavior?**

This PR adds a new flag option that generates a new entry in the <ClCompile> group for each configuration.

**Anything else we should know?**

Nope.

**Did you check all the boxes?**

- [x] Focus on a single fix or feature; remove any unrelated formatting or code changes
- [x] Add unit tests showing fix or feature works; all tests pass
- [x] Mention any [related issues](https://github.com/premake/premake-core/issues) (put `closes #XXXX` in comment to auto-close issue when PR is merged)
- [x] Follow our [coding conventions](https://github.com/premake/premake-core/blob/master/CONTRIBUTING.md#coding-conventions)
- [x] Minimize the number of commits
- [x] Align [documentation](https://github.com/premake/premake-core/tree/master/website) to your changes

*You can now [support Premake on our OpenCollective](https://opencollective.com/premake). Your contributions help us spend more time responding to requests like these!*

### CODE REVIEW

**Comment by Jarod42** [comment]
Is there a equivalent flag for other IDEs which might use msvc as compiler?

**Comment by LoneBoco** [comment]
> Is there a equivalent (command-line) flag for other IDEs which might use msvc as compiler?

No.  This isn't a command-line flag for the compiler.  This controls how the Visual Studio IDE scans for module dependencies and how it determines which files to compile.

This is related to using the `import` command.  By default, Visual Studio will only look at .ixx modules to build the dependency graph.  If it encounters something like `import <vector>;`, it will build that file as a header unit (.ifc), including any other headers it might `import`.

But say your main.cpp file had `import <iostream>;` and no .ixx files in your project tried to import the iostream header.  This would cause a build failure as iostream was never compiled into a header unit.

`Error C7612 could not find header unit for 'C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.30.30705\include\iostream'    MyProject    MyProject\src\main.cpp`

By enabling the `Scan Modules for Module Dependencies` flag, Visual Studio would have also scanned .cpp files and saw that we are trying to import iostream, so it would have built iostream as a header unit.  Then, when we compile, the build would be successful.

Other IDEs which use Visual Studio as a compiler would probably have their own way of determining which files to build.  They might scan .cpp files on their own, or maybe have a setting where you list additional files to compile.

The solutions given by Microsoft are:
1. Enable the `Scan Sources for Module Dependencies` option to scan every source file (increasing compile times).
2. Build a static shared library that imports header files, giving anything that links to it access to them.
3. Manually specifying that a .h file should be compiled as a header unit.  See PR #1570 for supporting CompileAs options.
4. Enable experimental STL modules and import them.

**Review by samsinsane — CHANGES_REQUESTED** [review]
We don't add new flags, please create an API instead.

**Comment by LoneBoco** [comment]
Okay.  It has been converted to an API call now.

File: modules/vstudio/vs2010_vcxproj.lua [hunk 1]
```diff
@@ -2873,6 +2874,19 @@
 	end
 
 
+	function m.scanSourceForModuleDependencies(cfg)
+		if _ACTION >= "vs2019" then
+			if cfg.scanformoduledependencies ~= nil then
```
Comment (nickclark2016) [inline modules/vstudio/vs2010_vcxproj.lua:2879]: So if the value isn't specified in the premake project file, we just use the IDE default value?

File: website/docs/scanformoduledependencies.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
+Enables the `Scan Sources for Module Dependencies` option for Visual Studio projects.
+
+```lua
+scanformoduledependencies "value"
+```
+If no value is set for a configuration, Visual Studio will default the option to `No`.
```
Comment (nickclark2016) [inline website/docs/scanformoduledependencies.md:6]: I don't think we need to specify this.  It should be noted that this is the default IDE value, not `No` based on my previous comment.

File: modules/vstudio/vs2010_vcxproj.lua (hunk 1 above)
Comment (LoneBoco) [reply modules/vstudio/vs2010_vcxproj.lua:2879]: Yes.  If the option isn't used then nothing gets written to the project file and the IDE's default is used.

File: website/docs/scanformoduledependencies.md (hunk 2 above)
Comment (LoneBoco) [reply website/docs/scanformoduledependencies.md]: I've removed the line.

**Review by nickclark2016 — APPROVED** [review]

**Pull request merged**